File: frontend/src/api/distribution_settlement.ts
import api from "./client";

/**
 * 신사업유통 정산 (자금 흐름) API 클라이언트 (T9 Phase F-C).
 *
 * 백엔드: `app/routers/distribution_settlement.py`
 *
 * 모든 GET 엔드포인트는 admin 권한 필요. 기간 필터(`from`/`to`)는 옵션이며
 * `YYYY-MM-DD` 문자열로 전송. `company_label` 은 옵션(미지정 시 전체).
 *
 * 데이터 의미:
 * - 매입: kr_purchase + vn_inventory_move + vn_sales_completed
 * - 입금요청: 자동계산 3종 합계 (KR매입×40% + VN재고×30% + VN매출×30%)
 * - 실 입금: account_deposit + cash_deposit
 * - 외상잔고: 시트 정의 — kr_purchase - 실 입금
 * - 이행률: 실 입금 / 입금요청 (0.0~1.0)
 */

const BASE = "/api/distribution/settlement";

export interface CashFlowItem {
  period_label: string;
  period_start: string; // ISO YYYY-MM-DD
  period_end: string; // ISO YYYY-MM-DD
  company_label: string;
  kr_purchase: number;
  vn_inventory_move: number;
  vn_sales_completed: number;
  kr_purchase_deposit_req: number;
  vn_inventory_deposit_req: number;
  vn_sales_deposit_req: number;
  deposit_req_total: number;
  account_deposit: number;
  cash_deposit: number;
  deposit_total: number;
  outstanding_balance: number;
  fulfillment_rate: number; // 0.0 ~ 1.0
}

export interface SettlementSummary {
  company_count: number;
  period_count: number;
  total_kr_purchase: number;
  total_vn_inventory_move: number;
  total_vn_sales: number;
  total_deposit_req: number;
  total_deposit_received: number;
  total_outstanding: number;
  fulfillment_rate: number; // 0.0 ~ 1.0
  latest_period_label: string | null;
}

export interface ByCompanyItem {
  company_label: string;
  period_count: number;
  total_kr_purchase: number;
  total_deposit_req: number;
  total_deposit_received: number;
  total_outstanding: number;
  fulfillment_rate: number; // 0.0 ~ 1.0
}

interface SettlementParams {
  from?: string;
  to?: string;
  company_label?: string;
}

function buildParams(params?: {
  from?: string;
  to?: string;
  company_label?: string;
}): SettlementParams {
  const out: SettlementParams = {};
  if (params?.from) out.from = params.from;
  if (params?.to) out.to = params.to;
  if (params?.company_label) out.company_label = params.company_label;
  return out;
}

export async function getCashFlow(params?: {
  from?: string;
  to?: string;
  company_label?: string;
}): Promise<CashFlowItem[]> {
  const res = await api.get<CashFlowItem[]>(`${BASE}/cash-flow`, {
    params: buildParams(params),
  });
  return res.data;
}

export async function getSettlementSummary(params?: {
  from?: string;
  to?: string;
  company_label?: string;
}): Promise<SettlementSummary> {
  const res = await api.get<SettlementSummary>(`${BASE}/summary`, {
    params: buildParams(params),
  });
  return res.data;
}

export async function getByCompany(params?: {
  from?: string;
  to?: string;
}): Promise<ByCompanyItem[]> {
  const res = await api.get<ByCompanyItem[]>(`${BASE}/by-company`, {
    params: buildParams(params),
  });
  return res.data;
}

export async function listSettlementCompanies(): Promise<string[]> {
  const res = await api.get<{ items: string[] }>(`${BASE}/companies`);
  return res.data.items;
}

File: backend/app/schemas/distribution_settlement.py
"""신사업유통 정산 (자금 흐름) Pydantic 응답 스키마 (T9 Phase F-C).

라우터: ``app/routers/distribution_settlement.py``
서비스: ``app/services/distribution/settlement_service.py``

직렬화 정책:
- 모든 금액은 ``float`` (서비스 단에서 Decimal → float 변환 완료).
- 카운트는 ``int``.
- ``fulfillment_rate`` 는 0.0~1.0.
- 날짜는 ``date`` (FastAPI 가 ISO YYYY-MM-DD 직렬화).
"""
from __future__ import annotations

from datetime import date

from pydantic import BaseModel, Field


class CashFlowItem(BaseModel):
    """주차별 정산 1행 — period_start 오름차순으로 응답.

    1주차 × 1회사 = 1행. 매입 3종 + 입금요청 3종 + 실 입금 2종 + 외상잔고 + 이행률.
    """

    period_label: str
    period_start: date
    period_end: date
    company_label: str
    kr_purchase: float
    vn_inventory_move: float
    vn_sales_completed: float
    kr_purchase_deposit_req: float
    vn_inventory_deposit_req: float
    vn_sales_deposit_req: float
    deposit_req_total: float
    account_deposit: float
    cash_deposit: float
    deposit_total: float
    outstanding_balance: float
    fulfillment_rate: float = Field(ge=0.0)


class SettlementSummary(BaseModel):
    """정산 요약 (KPI 카드용)."""

    company_count: int
    period_count: int
    total_kr_purchase: float
    total_vn_inventory_move: float
    total_vn_sales: float
    total_deposit_req: float
    total_deposit_received: float
    total_outstanding: float
    fulfillment_rate: float = Field(ge=0.0)
    latest_period_label: str | None = None


class ByCompanyItem(BaseModel):
    """회사별 합계 1행 (회사 비교 테이블).

    DISTRIBUTION_COMPANIES 4개 회사 모두 포함 — 데이터 없는 회사도 0 값으로 표시.
    """

    company_label: str
    period_count: int
    total_kr_purchase: float
    total_deposit_req: float
    total_deposit_received: float
    total_outstanding: float
    fulfillment_rate: float = Field(ge=0.0)


class CompaniesList(BaseModel):
    """회사 목록 응답 (Select 옵션 채움용)."""

    items: list[str]
